File: app/src/main/java/com/example/fondodepantalla/FragmentosAdministrador/AsistenciaManager.java
package com.example.fondodepantalla.FragmentosAdministrador;

import androidx.annotation.Nullable; // 🌟 AÑADIR ESTA IMPORTACIÓN

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

public class AsistenciaManager {

    private final FirebaseFirestore db;
    private final String uid;

    // 🌟 NUEVA INTERFAZ: Permite que el listener de Kotlin reciba 'null' en caso de error.
    public interface AsistenciaResultListener {
        void onResult(@Nullable DocumentSnapshot snapshot);
    }

    public AsistenciaManager(FirebaseFirestore db, String uid) {
        this.db = db;
        this.uid = uid;
    }

    // Registrar entrada (Sin cambios)
    public void registrarEntrada(Map<String, Object> data, OnSuccessListener<Void> onSuccess, OnFailureListener onFailure) {
        String fecha = data.get("fecha").toString();
        // NOTA: Usas "asistencias" -> uid -> "asistencias". Es poco común pero se respeta tu estructura.
        db.collection("asistencias")
                .document(uid)
                .collection("asistencias")
                .document(fecha)
                .set(data)
                .addOnSuccessListener(onSuccess)
                .addOnFailureListener(onFailure);
    }

    // 🌟 NUEVO MÉTODO: Reemplaza a 'registrarSalida'
    // Acepta un 'Map' para actualizar el documento, que es lo que el Kotlin envía.
    public void actualizarAsistencia(String fecha, Map<String, Object> data, OnSuccessListener<Void> onSuccess, OnFailureListener onFailure) {
        DocumentReference ref = db.collection("asistencias")
                .document(uid)
                .collection("asistencias")
                .document(fecha);

        ref.update(data) // Actualiza usando el Map
                .addOnSuccessListener(onSuccess)
                .addOnFailureListener(onFailure);
    }

    // 🌟 MÉTODO ANTIGUO ELIMINADO
    // public void registrarSalida(...) { ... }

    // 🌟 MÉTODO MODIFICADO: Usa la nueva interfaz 'AsistenciaResultListener'
    public void verificarAsistenciaHoy(String fecha, AsistenciaResultListener listener) {
        DocumentReference ref = db.collection("asistencias")
                .document(uid)
                .collection("asistencias")
                .document(fecha);

        ref.get()
                .addOnSuccessListener(snapshot -> listener.onResult(snapshot)) // Devuelve el snapshot en éxito
                .addOnFailureListener(e -> listener.onResult(null)); // Devuelve null en fallo
    }


    // 🌟 MÉTODO CORREGIDO: Añade comprobación de 'null'
    public void actualizarResumenSemanal(String fecha) {
        Calendar cal = Calendar.getInstance();
        int semana = cal.get(Calendar.WEEK_OF_YEAR);
        int año = cal.get(Calendar.YEAR);
        String idSemana = año + "-W" + semana;

        DocumentReference refResumen = db.collection("asistencias")
                .document(uid)
                .collection("resumenSemanal")
                .document(idSemana);

        // Usa el 'verificarAsistenciaHoy' modificado
        verificarAsistenciaHoy(fecha, snapshot -> {

            // 🌟 COMPROBACIÓN DE NULOS: Evita 'NullPointerException' si el get() falla
            if (snapshot == null || !snapshot.exists()) {
                // No se pudo obtener el documento, no se puede actualizar el resumen
                return;
            }

            String estado = snapshot.getString("estado");

            db.runTransaction(transaction -> {
                DocumentSnapshot snap = transaction.get(refResumen);
                long asistencias = snap.exists() && snap.contains("asistencias") ? snap.getLong("asistencias") : 0;
                long retardos = snap.exists() && snap.contains("retardos") ? snap.getLong("retardos") : 0;

                Map<String, Object> nuevosDatos = new HashMap<>();

                // 🌟 Añadida comprobación para 'sitio_cliente'
                if ("retardo".equals(estado)) {
                    nuevosDatos.put("retardos", retardos + 1);
                    // No sumamos asistencia si ya existe el documento
                } else if ("puntual".equals(estado) || "sitio_cliente".equals(estado)) {
                    nuevosDatos.put("asistencias", asistencias + 1);
                }

                // Si el estado es 'retardo', solo actualizamos 'retardos'
                if ("retardo".equals(estado)) {
                    transaction.update(refResumen, "retardos", retardos + 1);
                }
                // Si es la primera asistencia (puntual o cliente), creamos/actualizamos ambos
                else if (("puntual".equals(estado) || "sitio_cliente".equals(estado)) && !snap.exists()) {
                    nuevosDatos.put("asistencias", asistencias + 1);
                    nuevosDatos.put("retardos", retardos); // Asegura que 'retardos' exista
                    transaction.set(refResumen, nuevosDatos);
                }

                // Esta lógica es compleja. Una versión más simple sería:
                /*
                Map<String, Object> datos = new HashMap<>();
                long asistencias = snap.exists() ? snap.getLong("asistencias") : 0;
                long retardos = snap.exists() ? snap.getLong("retardos") : 0;

                if ("retardo".equals(estado)) {
                    retardos++;
                }
                // Contamos 'sitio_cliente' como asistencia
                if ("puntual".equals(estado) || "sitio_cliente".equals(estado)) {
                    asistencias++;
                }
                datos.put("asistencias", asistencias);
                datos.put("retardos", retardos);

                // .set(datos, SetOptions.merge()) es más seguro
                transaction.set(refResumen, datos, com.google.firebase.firestore.SetOptions.merge());
                */

                return null;
            });
        });
    }

    // 🌟 MÉTODO ANTIGUO ELIMINADO
    // private String dataHora() { ... }
    // (Ya no es necesario, la hora se genera en Kotlin)
}

File: app/src/main/java/com/example/fondodepantalla/data/AsistenciaManager.kt
package com.example.fondodepantalla.data

import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions // 2. Importar SetOptions para 'merge'
import java.util.Calendar

// 3. Constructor primario de Kotlin. 'db' y 'uid' se vuelven propiedades
// privadas de la clase automáticamente.
class AsistenciaManager(
    private val db: FirebaseFirestore,
    private val uid: String
) {

    // 4. OnSuccessListener y OnFailureListener reemplazados por lambdas:
    // () -> Unit  (no recibe nada, no devuelve nada)
    // (Exception) -> Unit (recibe una Excepción, no devuelve nada)
    fun registrarEntrada(
        data: Map<String, Any>, // 'Object' de Java es 'Any' en Kotlin
        onSuccess: () -> Unit,
        onFailure: (Exception) -> Unit
    ) {
        val fecha = data["fecha"].toString() // Acceso al mapa en Kotlin
        db.collection("asistencias")
            .document(uid)
            .collection("asistencias")
            .document(fecha)
            .set(data)
            .addOnSuccessListener { onSuccess() }
            .addOnFailureListener(onFailure) // Pasa 'it' (la excepción) implícitamente
    }

    fun actualizarAsistencia(
        fecha: String,
        data: Map<String, Any>,
        onSuccess: () -> Unit,
        onFailure: (Exception) -> Unit
    ) {
        db.collection("asistencias")
            .document(uid)
            .collection("asistencias")
            .document(fecha)
            .update(data)
            .addOnSuccessListener { onSuccess() }
            .addOnFailureListener(onFailure)
    }

    // 5. 'AsistenciaResultListener' reemplazado por (DocumentSnapshot?) -> Unit
    // Un lambda que acepta un DocumentSnapshot nulable.
    fun verificarAsistenciaHoy(
        fecha: String,
        listener: (DocumentSnapshot?) -> Unit
    ) {
        db.collection("asistencias")
            .document(uid)
            .collection("asistencias")
            .document(fecha)
            .get()
            .addOnSuccessListener { snapshot -> listener(snapshot) }
            .addOnFailureListener { listener(null) } // Devuelve 'null' en caso de fallo
    }

    fun actualizarResumenSemanal(fecha: String) {
        val cal = Calendar.getInstance()
        val semana = cal.get(Calendar.WEEK_OF_YEAR)
        val año = cal.get(Calendar.YEAR)
        // 6. Interpolación de strings en Kotlin (más limpio)
        val idSemana = "$año-W$semana"

        val refResumen = db.collection("asistencias")
            .document(uid)
            .collection("resumenSemanal")
            .document(idSemana)

        // 7. 'verificarAsistenciaHoy' ahora usa un bloque lambda
        verificarAsistenciaHoy(fecha) { snapshot ->

            // 8. Verificación de nulos segura
            if (snapshot == null || !snapshot.exists()) {
                return@verificarAsistenciaHoy // Sale solo de este lambda
            }

            val estado = snapshot.getString("estado")

            db.runTransaction { transaction ->
                val snap = transaction.get(refResumen)

                // 9. Lógica de transacción simplificada (como sugeriste en tu comentario de Java)
                // '?: 0L' es el "operador Elvis": si lo de la izquierda es nulo, usa '0L'.
                var asistencias = if (snap.exists()) snap.getLong("asistencias") ?: 0L else 0L
                var retardos = if (snap.exists()) snap.getLong("retardos") ?: 0L else 0L

                if (estado == "retardo") {
                    retardos++
                }
                // Contamos 'sitio_cliente' como asistencia
                if (estado == "puntual" || estado == "sitio_cliente") {
                    asistencias++
                }

                // 10. Creación de mapas en Kotlin
                val datos = mapOf(
                    "asistencias" to asistencias,
                    "retardos" to retardos
                )

                // .set con SetOptions.merge() es la forma más segura de actualizar
                // o crear un documento sin sobrescribir campos no intencionados.
                transaction.set(refResumen, datos, SetOptions.merge())

                // Requerido por la transacción de Firestore en Kotlin
                null
            }
        }
    }
}